vshard/replicaset.lua

```diff
@@ -577,13 +577,33 @@ local function replicaset_template_multicallro(prefer_replica, balance)
     local function pick_next_replica(replicaset, now)
         local r
         local master = replicaset.master
+
         if balance then
+            local prefered_zone = replicaset.priority_list[1].zone
             local i = #replicaset.priority_list
             while i > 0 do
                 r = replicaset_balance_replica(replicaset)
                 i = i - 1
                 if r:is_connected() and (not prefer_replica or r ~= master) and
-                   replica_check_backoff(r, now) then
+                    replica_check_backoff(r, now)
+                then
+                    -- Pick a replica according prefered zone (highest priority replica zone) in round-robin manner
+                    if balance == "prefer_zone" and prefered_zone then
+                        local cbi = replicaset.balance_i
+                        local nr = replicaset_balance_replica(replicaset)
+
+                        if prefered_zone and r.zone and r.zone == prefered_zone
+                            and nr.zone and nr.zone ~= prefered_zone
+                            and (not prefer_replica or nr ~= master)
+                        then
+                            -- Reset cursor to the main position if next replica is in different zone.
+                            replicaset.balance_i = 1
+                        else
+                            -- Restore rr-cursor position.
+                            replicaset.balance_i = cbi
+                        end
+                    end
+
```

Comment:
You should strive to keep the code indentation to the minimum. For example, here it is easily achievable by just handling simple cases first and doing early return. Then you can handle the zone balancing with -4 spaces indentation:
```Diff
                 if r:is_connected() and (not prefer_replica or r ~= master) and
-                    replica_check_backoff(r, now)
-                then
+                   replica_check_backoff(r, now) then
+                    if balance ~= 'prefered_zone' or not prefered_zone then
+                        return r
+                    end
                     -- Pick a replica according prefered zone (highest priority replica zone) in round-robin manner
-                    if balance == "prefer_zone" and prefered_zone then
-                        local cbi = replicaset.balance_i
-                        local nr = replicaset_balance_replica(replicaset)
+                    local cbi = replicaset.balance_i
+                    local nr = replicaset_balance_replica(replicaset)
 
-                        if prefered_zone and r.zone and r.zone == prefered_zone
-                            and nr.zone and nr.zone ~= prefered_zone
-                            and (not prefer_replica or nr ~= master)
-                        then
-                            -- Reset cursor to the main position if next replica is in different zone.
-                            replicaset.balance_i = 1
-                        else
-                            -- Restore rr-cursor position.
-                            replicaset.balance_i = cbi
-                        end
+                    if prefered_zone and r.zone and r.zone == prefered_zone
+                        and nr.zone and nr.zone ~= prefered_zone
+                        and (not prefer_replica or nr ~= master)
+                    then
+                        -- Reset cursor to the main position if next replica is in different zone.
+                        replicaset.balance_i = 1
+                    else
+                        -- Restore rr-cursor position.
+                        replicaset.balance_i = cbi
                     end
-
```